common/beerocks/bcl/source/son/son_wireless_utils.cpp

```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+                   << " Mbps)";
+
+        estimation_normal.status   = ESTIMATION_SUCCESS;
+        estimation_normal.tx_power = is_5ghz ? phy_rate_table[max_ant_mode][max_mcs].tx_power_5
+                                             : phy_rate_table[max_ant_mode][max_mcs].tx_power_2_4;
+        estimation_normal.rssi =
+            int(ceil(phy_rate_table[max_ant_mode][max_mcs].bw_values[max_bw].rssi / 10.0));
+
+        LOG(DEBUG) << "Return maximal estimation values | tx_power:" << estimation_normal.tx_power
+                   << " | RSSI:" << estimation_normal.rssi;
+
+        return estimation_normal;
+    }
+
+    for (int ant_mode = max_ant_mode; ant_mode > -1; ant_mode--) { // filter by ant_mode
+        for (int bw = max_bw; bw > -1; bw--) {                     // filter by max_bw
             for (int mcs = max_mcs; mcs > -1; mcs--) {             // filter by mcs
 
-                estimated_ul_rssi_lut = phy_rate_table[ant_mode][mcs].bw_values[bw].rssi;
-                estimated_ul_rssi_lut_delta =
-                    (int)std::abs((float)(ul_rssi_lut - estimated_ul_rssi_lut));
+                estimated_ul_rssi_lut       = phy_rate_table[ant_mode][mcs].bw_values[bw].rssi;
+                estimated_ul_rssi_lut_delta = std::abs(ul_rssi_lut - estimated_ul_rssi_lut);
 
-                // same rate && min rssi_delta
-                if ((sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate) ||
-                    (sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate)) {
+                auto gi_long_rate  = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate;
+                auto gi_short_rate = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate;
+
+                // Check if the current rate is between the gi_long_rate to the gi_short_rate
+                if ((sta_phy_tx_rate_100kb >= gi_long_rate) &&
+                    (sta_phy_tx_rate_100kb <= gi_short_rate)) {
+                    // phyrate is in range - use table values
                     if (estimated_ul_rssi_lut_delta <= estimated_ul_rssi_lut_delta_min) {
                         estimated_ul_rssi_lut_delta_min = estimated_ul_rssi_lut_delta;
-                        estimation.tx_power = is_5ghz ? phy_rate_table[ant_mode][mcs].tx_power_5
-                                                      : phy_rate_table[ant_mode][mcs].tx_power_2_4;
-                        estimation.rssi =
-                            int(float(estimated_ul_rssi_lut) / 10.0 + 0.5); //round value
+                        estimation_normal.tx_power =
+                            is_5ghz ? phy_rate_table[ant_mode][mcs].tx_power_5
+                                    : phy_rate_table[ant_mode][mcs].tx_power_2_4;
+                        estimation_normal.rssi = int(ceil(estimated_ul_rssi_lut / 10.0));
```

Comment:
It is the same result. I found it stupid to add 0.5 and then cast it to `int` so the number will be rounded up.

Reply:
The original code was `int(float(est) / 10.0 + 0.5)`. Sinds int() rounds down, this rounds down if the last decimal is smaller than 5 and up if it is larger than 5. The new code is `int(ceil(est / 10.0))` which always rounds up.

See this test code:

```
#include <iostream>
#include <cmath>

int main() {
  int est = 14;
  std::cout << "div: " << (float(est) / 10.0) << " int: " << int(float(est)/10.0 + 0.5) << std::endl;
  std::cout << "div: " << (est / 10.0) << " ceil: " << ceil(est/10.0) << " int: " << int(ceil(est/10.0)) << std::endl;
}
```

output:

```
div: 1.4 int: 1
div: 1.4 ceil: 2 int: 2
```

Reply:
I thought it will round to the nearest integer value.
Anyway, according to the comments. the one who coded it meant to round it up. So in that case, it fixes it.